watch/watch.go

```diff
@@ -0,0 +1,360 @@
+	w := &Watcher{
+		id:       id,
+		store:    s,
+		compiler: c,
+		ctx:      ctx,
+
+		queryWatch: map[string]queryNotify{},
+		dataWatch:  map[string]map[signal]struct{}{},
+	}
+	w.indexRules(c)
+	return w
+}
+
+func (w *Watcher) query(query, id string, out chan<- Event) error {
+	w.l.Lock()
+	defer w.l.Unlock()
+
+	if w.closed {
+		return errors.New("watcher is closed")
+	}
+
+	if _, ok := w.queryWatch[id]; ok {
+		return fmt.Errorf("already a watch with id %s", id)
+	}
+
+	parsed, err := ast.ParseBody(query)
+	if err != nil {
+		return err
+	}
+
+	compiled, err := w.compiler.QueryCompiler().Compile(parsed)
+	if err != nil {
+		return err
+	}
+	refs := w.extractBaseRefs(compiled)
+
+	notify := make(signal, 1)
+	done := make(signal)
+	ack := make(signal)
+	go w.deliver(query, id, out, notify, done, ack)
+
+	w.queryWatch[id] = queryNotify{query, out, notify, done, ack}
+	for _, r := range refs {
+		s := r.String()
+		if _, ok := w.dataWatch[s]; !ok {
+			w.dataWatch[s] = map[signal]struct{}{}
+		}
+		w.dataWatch[s][notify] = struct{}{}
+	}
+
+	// Send one notification when we start (like first query).
+	select {
+	case notify <- struct{}{}:
+	default:
+	}
+
+	return nil
+}
+
+func (w *Watcher) deliver(query string, id string, out chan<- Event, notify, done, ack signal) {
+	defer close(ack)
+	for {
+		select {
+		case <-notify:
+			result, err := w.evalQuery(query)
+			out <- Event{ID: id, Value: result, Error: err}
+		case <-done:
+			return
+		}
+	}
+}
+
+// endQuery assumes the lock on the Watcher is held.
+func (w *Watcher) endQuery(id string, migrating bool) error {
+	notify, err := w.deleteQuery(id)
+	if err != nil {
+		return err
+	}
+	go cleanupQuery(notify, migrating)
+	return nil
+}
+
+func (w *Watcher) deleteQuery(id string) (queryNotify, error) {
+	notify, ok := w.queryWatch[id]
+	if !ok {
+		return queryNotify{}, fmt.Errorf("no query watch with id %s", id)
+	}
+	delete(w.queryWatch, id)
+
+	for _, notifiers := range w.dataWatch {
+		delete(notifiers, notify.notify)
+	}
+	return notify, nil
+}
+
+// close assumes that the Watcher is locked. It will not unlock it.
+func (w *Watcher) close(txn storage.Transaction, migrating bool) {
+	w.store.Unregister(w.ctx, txn, w.id)
+	for id := range w.queryWatch {
+		w.endQuery(id, migrating)
+	}
+}
+
+func (w *Watcher) notify(ctx context.Context, txn storage.Transaction, event storage.TriggerEvent) {
+	w.l.Lock()
+	defer w.l.Unlock()
+
+	if w.closed {
+		return
+	}
+
+	for _, d := range event.Data {
+		r := d.Path.Ref(ast.DefaultRootDocument)
+		for path, notifiers := range w.dataWatch {
+			if r.HasPrefix(ast.MustParseRef(path)) {
+				sendNotifications(notifiers)
+			}
+		}
+	}
+}
+
+func (w *Watcher) indexRules(c *ast.Compiler) {
+	w.rules = nil
+	for _, mod := range c.Modules {
+		ast.WalkRules(mod, func(r *ast.Rule) bool {
+			w.rules = append(w.rules, r)
+			return false
+		})
+	}
+}
+
+func (w *Watcher) lookupRule(r ast.Ref) (rules []*ast.Rule) {
+	var prefix, postfix bool
```

Comment:
Is there a reason why the existing lookup functions on the compiler are insufficient here?